File: ciDoBerita.py
import time 
#Saya sudah coba buat menggunakan fungsi delay sendiri namun ternyata boros CPU sehingga saya memilih menggunakan library

class Node:
    def __init__(self, data):
        self.data = data
        self.next = self
        self.prev = self

class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        tail = self.head.prev
        tail.next = new_node
        new_node.prev = tail
        new_node.next = self.head
        self.head.prev = new_node

    def delete(self, pos):
        if self.head is None:
            return
        temp = self.head
        count = 1
        while True:
            if count == pos:
                if temp.next == temp:
                    self.head = None
                    return
                temp.prev.next = temp.next
                temp.next.prev = temp.prev
                if temp == self.head:
                    self.head = temp.next
                return
            temp = temp.next
            count += 1
            if temp == self.head:
                print("Position out of range.")
                return

    def display_forward(self):
        if self.head is None:
            print("List kosong.")
            return
        temp = self.head
        first = True
        while True:
            if not first:
                time.sleep(3)
            print(temp.data)
            first = False
            temp = temp.next
            if temp == self.head:
                break

    def display_backward(self):
        if self.head is None:
            print("List kosong.")
            return
        temp = self.head.prev
        first = True
        while True:
            if not first:
                time.sleep(3)
            print(temp.data)
            first = False
            temp = temp.prev
            if temp == self.head.prev:
                break

    def display_at(self, pos):
        if self.head is None:
            print("List kosong.")
            return
        temp = self.head
        count = 1
        while True:
            if count == pos:
                print(temp.data)
                return
            temp = temp.next
            count += 1
            if temp == self.head:
                print("Position out of range.")
                return

def main():
    lst = CircularDoublyLinkedList()
    while True:
        print("1.Insert berita")
        print("2.Hapus berita")
        print("3.Tampilkan forward")
        print("4.Tampilkan backward")
        print("5.Tampil berita tertentu")
        print("6.Exit")
        print("Pilih menu: ", end="")
        choice = int(input())
        if choice == 1:
            text = input("Masukkan berita: ")
            lst.insert(text)
        elif choice == 2:
            pos = int(input("Masukkan posisi: "))
            lst.delete(pos)
        elif choice == 3:
            lst.display_forward()
        elif choice == 4:
            lst.display_backward()
        elif choice == 5:
            p = int(input("Masukkan posisi: "))
            lst.display_at(p)
        elif choice == 6:
            print("Keluar...")
            break
        else:
            print("Pilihan tidak valid.")

if __name__ == "__main__":
    main()

File: ciDoBerita.java
import java.util.Scanner;

class Node {
    String data;
    Node next, prev;
    Node(String data) {
        this.data = data;
        this.next = this.prev = null;
    }
}

class CircularDoublyLinkedList {
    Node head = null;

    void insert(String data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            head.next = head.prev = head;
            return;
        }
        Node tail = head.prev;
        tail.next = newNode;
        newNode.prev = tail;
        newNode.next = head;
        head.prev = newNode;
    }

    void delete(int pos) {
        if (head == null) return;
        Node temp = head;
        int count = 1;
        do {
            if (count == pos) {
                if (temp.next == temp) {
                    head = null;
                    return;
                }
                temp.prev.next = temp.next;
                temp.next.prev = temp.prev;
                if (temp == head) head = temp.next;
                return;
            }
            temp = temp.next;
            count++;
        } while (temp != head);
        System.out.println("Position out of range.");
    }

    void displayForward() throws InterruptedException {
        if (head == null) { System.out.println("List kosong."); return; }
        Node temp = head;
        do {
            System.out.println(temp.data);
            temp = temp.next;
            if (temp != head) Thread.sleep(3000);
        } while (temp != head);
    }

    // FIX: was stopping before printing head node (off-by-one in loop termination)
    void displayBackward() throws InterruptedException {
        if (head == null) { System.out.println("List kosong."); return; }
        Node temp = head.prev; // start at tail
        do {
            System.out.println(temp.data);
            temp = temp.prev;
            if (temp != head.prev) Thread.sleep(3000);
        } while (temp != head.prev); // stop after full circle
    }

    void displayAt(int pos) {
        if (head == null) { System.out.println("List kosong."); return; }
        Node temp = head;
        int count = 1;
        do {
            if (count == pos) {
                System.out.println(temp.data);
                return;
            }
            temp = temp.next;
            count++;
        } while (temp != head);
        System.out.println("Position out of range.");
    }
}

public class ciDoBerita {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        CircularDoublyLinkedList list = new CircularDoublyLinkedList();
        int choice;
        do {
            System.out.println("1.Insert berita");
            System.out.println("2.Hapus berita");
            System.out.println("3.Tampilkan forward");
            System.out.println("4.Tampilkan backward");
            System.out.println("5.Tampil berita tertentu");
            System.out.println("6.Exit");
            System.out.print("Pilih: ");
            choice = sc.nextInt();
            sc.nextLine();
            switch (choice) {
                case 1:
                    System.out.print("Masukkan berita: ");
                    String text = sc.nextLine();
                    list.insert(text);
                    break;
                case 2:
                    System.out.print("Masukkan posisi: ");
                    int pos = sc.nextInt();
                    list.delete(pos);
                    break;
                case 3:
                    list.displayForward();
                    break;
                case 4:
                    list.displayBackward();
                    break;
                case 5:
                    System.out.print("Masukkan posisi: ");
                    int p = sc.nextInt();
                    list.displayAt(p);
                    break;
                case 6:
                    System.out.println("Keluar...");
                    break;
                default:
                    System.out.println("Pilihan tidak valid.");
            }
        } while (choice != 6);
        sc.close();
    }
}